File: OneDrive/Cerberus/Cerberus/js/spider.js
const { URL } = require("url");

function arg(name, def) {
  const idx = process.argv.indexOf(name);
  if (idx !== -1 && idx + 1 < process.argv.length) return process.argv[idx + 1];
  return def;
}

async function fetchText(u, timeout) {
  const c = AbortController ? new AbortController() : null;
  const t = c ? setTimeout(() => c.abort(), timeout) : null;
  try {
    const r = await fetch(u, { redirect: "follow", signal: c ? c.signal : undefined });
    const ct = r.headers.get("content-type") || "";
    if (ct.includes("text/html") || ct === "") return await r.text();
    return "";
  } catch {
    return "";
  } finally {
    if (t) clearTimeout(t);
  }
}

function sameOrigin(base, u) {
  try {
    const a = new URL(base);
    const b = new URL(u, base);
    return a.protocol === b.protocol && a.host === b.host;
  } catch {
    return false;
  }
}

function abs(base, u) {
  try {
    return new URL(u, base).toString();
  } catch {
    return null;
  }
}

function parseHTML(base, html) {
  const links = new Set();
  const scripts = new Set();
  const forms = [];
  const api = new Set();
  const aRe = /<a[^>]+href=["']([^"'#]+)["']/gi;
  const sRe = /<script[^>]+src=["']([^"']+)["']/gi;
  const fRe = /<form[^>]*?(?:action=["']?([^"'>\s]*)[^>]*?)?(?:method=["']?([^"'>\s]*)[^>]*?)?>/gi;
  const iRe = /<input[^>]*name=["']([^"']+)["'][^>]*?(?:type=["']?([^"'>\s]*)[^>]*)?/gi;
  let m;
  while ((m = aRe.exec(html))) {
    const u = abs(base, m[1]);
    if (u && sameOrigin(base, u)) {
      links.add(u);
      if (u.includes("/api/") || u.endsWith(".json")) api.add(u);
    }
  }
  while ((m = sRe.exec(html))) {
    const u = abs(base, m[1]);
    if (u) scripts.add(u);
  }
  while ((m = fRe.exec(html))) {
    const action = abs(base, m[1] || "");
    const method = (m[2] || "GET").toUpperCase();
    const inputs = [];
    let mi;
    while ((mi = iRe.exec(html))) {
      inputs.push({ name: mi[1], type: mi[2] || "text" });
    }
    forms.push({ action: action || base, method, inputs });
  }
  return { links: Array.from(links), scripts: Array.from(scripts), forms, api_endpoints: Array.from(api) };
}

async function run(target, maxDepth, maxPages, timeout) {
  const visited = new Set();
  const pages = new Set();
  const forms = [];
  const scripts = new Set();
  const api = new Set();
  const q = [[target, 0]];
  while (q.length && pages.size < maxPages) {
    const [u, d] = q.shift();
    if (visited.has(u) || d > maxDepth) continue;
    visited.add(u);
    const html = await fetchText(u, timeout);
    if (!html) continue;
    pages.add(u);
    const parsed = parseHTML(u, html);
    parsed.forms.forEach(f => forms.push(f));
    parsed.scripts.forEach(s => scripts.add(s));
    parsed.api_endpoints.forEach(a => api.add(a));
    for (const l of parsed.links) {
      if (!visited.has(l)) q.push([l, d + 1]);
    }
  }
  return {
    pages: Array.from(pages).sort(),
    forms,
    scripts: Array.from(scripts).sort(),
    api_endpoints: Array.from(api).sort()
  };
}

async function main() {
  const target = arg("--target");
  if (!target) {
    console.error("usage: node spider.js --target <url> [--depth 2] [--limit 200] [--timeout 8000]");
    process.exit(1);
  }
  const depth = parseInt(arg("--depth", "2"), 10);
  const limit = parseInt(arg("--limit", "200"), 10);
  const timeout = parseInt(arg("--timeout", "8000"), 10);
  const data = await run(target, depth, limit, timeout);
  const out = JSON.stringify(data, null, 2);
  console.log(out);
}

main().catch(e => {
  console.error(String(e || "error"));
  process.exit(1);
});

File: OneDrive/Cerberus/Cerberus/recon/spider.py
import requests
from urllib.parse import urlparse, urljoin
from html.parser import HTMLParser
from collections import deque


class LinkFormParser(HTMLParser):
    def __init__(self, base_url):
        super().__init__()
        self.base_url = base_url
        self.links = set()
        self.forms = []
        self.scripts = set()
        self._current_form = None

    def handle_starttag(self, tag, attrs):
        attrs_dict = dict(attrs)
        if tag == "a":
            href = attrs_dict.get("href")
            if href:
                url = urljoin(self.base_url, href)
                self.links.add(url)
        elif tag == "form":
            action = attrs_dict.get("action") or ""
            method = attrs_dict.get("method") or "GET"
            self._current_form = {"action": urljoin(self.base_url, action), "method": method.upper(), "inputs": []}
        elif tag == "input":
            if self._current_form is not None:
                name = attrs_dict.get("name")
                input_type = attrs_dict.get("type") or "text"
                self._current_form["inputs"].append({"name": name, "type": input_type})
        elif tag == "script":
            src = attrs_dict.get("src")
            if src:
                url = urljoin(self.base_url, src)
                self.scripts.add(url)

    def handle_endtag(self, tag):
        if tag == "form" and self._current_form is not None:
            self.forms.append(self._current_form)
            self._current_form = None


class Spider:
    def __init__(self, target, max_depth=2, max_pages=200, timeout=8):
        self.target = target.rstrip("/")
        self.parsed = urlparse(self.target)
        self.base = f"{self.parsed.scheme}://{self.parsed.netloc}"
        self.max_depth = max_depth
        self.max_pages = max_pages
        self.timeout = timeout
        self.visited = set()
        self.pages = set()
        self.forms = []
        self.scripts = set()
        self.api_endpoints = set()

    def _same_origin(self, url):
        p = urlparse(url)
        return p.netloc == self.parsed.netloc and p.scheme in ("http", "https")

    def _fetch(self, url):
        try:
            r = requests.get(url, timeout=self.timeout, allow_redirects=True)
            ctype = r.headers.get("Content-Type", "")
            if "text/html" in ctype or ctype == "":
                return r.text
            return ""
        except requests.RequestException:
            return ""

    def _parse_page(self, url, html):
        parser = LinkFormParser(url)
        parser.feed(html)
        links = {l for l in parser.links if self._same_origin(l)}
        self.forms.extend(parser.forms)
        self.scripts |= parser.scripts
        for l in links:
            if "/api/" in l or l.endswith(".json"):
                self.api_endpoints.add(l)
        return links

    def run(self):
        queue = deque([(self.target, 0)])
        while queue and len(self.pages) < self.max_pages:
            url, depth = queue.popleft()
            if url in self.visited or depth > self.max_depth:
                continue
            self.visited.add(url)
            html = self._fetch(url)
            if not html:
                continue
            self.pages.add(url)
            links = self._parse_page(url, html)
            for l in links:
                if l not in self.visited:
                    queue.append((l, depth + 1))
        return {
            "pages": sorted(self.pages),
            "forms": self.forms,
            "scripts": sorted(self.scripts),
            "api_endpoints": sorted(self.api_endpoints),
        }
